### ISSUE TITLE
New lint: Give a warning when calling File::bytes, TcpStream::bytes(), etc

### ISSUE THREAD

**Issue by JonathanBrouwer** [rust-lang/rust-clippy#14087]
### What it does

There should be a lint that gives a warning when calling File::bytes(), TcpStream::bytes(), and maybe others. 
These functions (almost?) should never be called since File and TcpStream are unbuffered and will perform a syscall everytime .next() is called. This is also noted on the [docs page of the function](https://doc.rust-lang.org/stable/std/io/trait.Read.html#method.bytes).

I'm not sure an automatic fix will be possible, since if the bytes() is used only partially, the file is only read partially. Inserting the bufreader would change this behaviour.

### Advantage

- I measured the rewritten code to be ~1000x faster for large file

### Drawbacks

- Theoretically a user could write `file.bytes().next()`, then dropping the iterator. In this case, adding the BufReader would make no difference. I expect this to be a rare situation though. 

### Example

```rust
let file = File::open("./bytes.txt").unwrap();
let zero_bytes_count = file.bytes().map(|b| b.unwrap()).filter(|b| *b == 0).count();
```

Could be written as:

```rust
let file = BufReader::new(File::open("./bytes.txt").unwrap());
let zero_bytes_count = file.bytes().map(|b| b.unwrap()).filter(|b| *b == 0).count();
```

**Comment by JonathanBrouwer** [comment]
I asked for feedback on the zulip here as well https://rust-lang.zulipchat.com/#narrow/channel/257328-clippy/topic/Implementing.20a.20new.20clippy.20lint/near/496167132
I'll give implementing this a try
